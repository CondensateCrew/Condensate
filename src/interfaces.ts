// DATA INTERFACES
export interface UserLoginPosting {
  email: string,
  password: string
}

export interface UserLoginReceived {
  readonly id: number,
  readonly firstName: string,
  readonly lastName: string
}

export interface UserSignupPosting {
  email: string,
  password: string,
  firstName: string,
  lastName: string,
  id?: number
}

export interface Action {
  id: number,
  name: string
}

export interface Brainstorm {
  id: number,
  question: string,
  idea: string,
  action: string,
  isGenius: boolean,
  categories: Category[]
}

export interface Category {
  id: number,
  name: string
}

export interface Insight {
  id: number,
  question: string,
  answers: string[]
}

export interface WordContext {
  type: string,
  connectedWord: string[]
}

export interface RandomWordCollection {
  [key: string]: WordContext
}

export interface AppStore {
  user: UserLoginReceived,
  allBrainstorms: Brainstorm[],
  actions: Action[],
  categories: Category[],
  readonly randomWordCollections: RandomWordCollection[],
  readonly questionTemplates: string[],
  readonly secretSauce: string[],
  currentBrainstorm: Brainstorm,
  chosenWords: string[],
  insights: Insight[],
  query: string,
  filter: string,
  timeEnded: boolean,
  instructionsEnded: boolean
}

// ACTIONS OBJECTS

export interface ActionObject {
  type: string,
  [key: string]: any
}

export interface IAddUserAction {
  type: 'ADD_USER',
  user: UserLoginReceived
}

export interface ILogoutUserAction {
  type: 'LOG_OUT_USER'
}

export interface IAddAllActionsAction {
  type: 'ADD_ALL_ACTIONS',
  actions: Action[]
}

export interface IAddNewActionAction {
  type: 'ADD_NEW_ACTION',
  action: Action
}

export interface IAddAllBrainstormsAction {
  type: 'ADD_ALL_BRAINSTORMS',
  brainstorms: Brainstorm[]
}

export interface IChangeBrainstormAction {
  type: 'TOGGLE_BRAINSTORMS',
  id: number
}

export interface IAddNewBrainstormAction {
  type: 'ADD_BRAINSTORM',
  brainstorm: Brainstorm
}

export interface IDeleteBrainstormAction {
  type: 'DELETE_BRAINSTORM',
  id: number
}

export interface IAddAllCategoriesAction {
  type: 'ADD_ALL_CATEGORIES',
  categories: Category[]
}

export interface IAddNewCategoryAction {
  type: 'ADD_CATEGORY',
  category: Category
}

export interface IAddRandomWordCollectionsAction {
  type: 'ADD_ALL_COLLECTIONS',
  collections: RandomWordCollection[]
}

export interface IAddQuestionTemplatesAction {
  type: 'ADD_ALL_TEMPLATES',
  templates: string[]
}

export interface IAddSecretSauceAction {
  type: 'ADD_SECRETE_SAUSE',
  secretSauce: string[]
}

export interface IAddCurrentBrainstormAction {
  type: 'ADD_CURRENT_BRAINSTORM',
  currentBrainstorm: Brainstorm
}

export interface IRemoveCurrentBrainstormAction {
  type: 'REMOVE_CURRENT_BRAINSTORM'
}

export interface IAddChosenWordAction {
  type: 'ADD_WORD',
  chosenWord: string
}

export interface IRemoveChosenWordAction {
  type: 'REMOVE_WORD',
  chosenWord: string
}

export interface IAddInsightAction {
  type: 'ADD_INSIGHT',
  insight: Insight
}

export interface IAddQueryAction {
  type: 'ADD_QUERY',
  query: string
}

export interface IRemoveQueryAction {
  type: 'REMOVE_QUERY'
}

export interface IAddFilterAction {
  type: 'ADD_FILTER',
  filter: string
}

export interface IRemoveFilterAction {
  type: 'REMOVE_FILTER'
}

export interface IEndTimeAction {
  type: 'END_TIME'
}

export interface IReverseTimeAction {
  type: 'REVERSE_TIME'
}

export interface IEndInstructionsAction {
  type: 'END_INSTRUCTIONS'
}

export interface IReverseInstructionsAction {
  type: 'REVERSE_INSTRUCTIONS'
}
