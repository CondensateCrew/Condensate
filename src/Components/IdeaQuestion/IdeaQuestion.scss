@import '../../css/variables.scss';
@import '../../css/index.scss';

.idea-question-input {
  @extend %regular-input;
}
