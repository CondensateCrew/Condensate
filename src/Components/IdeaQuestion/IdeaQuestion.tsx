import React, { useState, useEffect } from 'react';
import './IdeaQuestion.scss';
import { IBrainstormForm } from '../../interfaces';

interface Props {
  formState: IBrainstormForm,
  setQuestion: (formState: IBrainstormForm) => void
};

const IdeaQuestion:React.FC<Props> = ({ formState, setQuestion }) => {
  const [ questionValue, setQuestionValue ] = useState<string>('');
  const updateQuestion = ():void => setQuestion({...formState, question: questionValue});
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setQuestionValue(e.target.value);
  };

  useEffect(updateQuestion, [questionValue]);
  
  useEffect(() => formState.reset 
  ? setQuestionValue('') 
  : setQuestionValue(questionValue)
  , [formState.reset, questionValue]);

  return (
    <input
      className='idea-question-input'
      value={questionValue}
      type='text'
      placeholder='Enter your idea here...'
      onChange={handleChange}/>
  );
}

export default IdeaQuestion;
