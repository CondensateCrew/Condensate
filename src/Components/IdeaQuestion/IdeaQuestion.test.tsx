/**
 * @jest-environment jsdom
 */

import React from 'react';
import { mount } from 'enzyme';
import { IBrainstormForm } from '../../interfaces';
import IdeaQuestion from './IdeaQuestion';

describe('IdeaQuestion Component', () => {
  let wrapper: any;

  let mockFormState: IBrainstormForm = {
    question: '',
    categories: [],
    action: {id: 1, action: 'create an app'},
    reset: false
  };
  let mockSetQuestion = jest.fn();

  beforeEach(() => {
    wrapper = mount(<IdeaQuestion
      formState={mockFormState}
      setQuestion={mockSetQuestion}/>);
  })

  afterEach(() => {
    jest.clearAllMocks();
  })
  it('should match the snapshot with all the data passed through', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call setQuestion after changes in input occur', () => {
    const mockEvent: Object = {target: {value: 'What is love?'}}
    wrapper.find('input').simulate('change', mockEvent);
    expect(mockSetQuestion).toHaveBeenCalled();
  });

  it('should update question when change events occur', () => {
    const mockEvent: Object = {target: {value: 'What is love?'}}
    wrapper.find('input').simulate('change', mockEvent);
    expect(wrapper.find('input').getDOMNode().value).toEqual('What is love?')
  });

  it('should reset its state when formState.reset is true', () => {
    let mockFormState: IBrainstormForm = {
      question: '',
      categories: [],
      action: {id: 1, action: 'create an app'},
      reset: true
    };
    wrapper = mount(<IdeaQuestion
      formState={mockFormState}
      setQuestion={mockSetQuestion}/>);


  const mockEvent: Object = {target: {value: 'What is love?'}}
  wrapper.find('input').simulate('change', mockEvent);
  expect(wrapper.find('input').getDOMNode().value).toEqual('')
  });
});
