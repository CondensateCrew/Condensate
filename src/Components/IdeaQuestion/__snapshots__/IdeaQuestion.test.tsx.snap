// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdeaQuestion Component should match the snapshot with all the data passed through 1`] = `
<IdeaQuestion
  formState={
    Object {
      "action": Object {
        "action": "create an app",
        "id": 1,
      },
      "categories": Array [],
      "question": "",
      "reset": false,
    }
  }
  setQuestion={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "action": Object {
              "action": "create an app",
              "id": 1,
            },
            "categories": Array [],
            "question": "",
            "reset": false,
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <input
    className="idea-question-input"
    onChange={[Function]}
    placeholder="Enter your idea here..."
    type="text"
    value=""
  />
</IdeaQuestion>
`;
